design/StableRegions-20230209.md

```diff
@@ -0,0 +1,172 @@
+control the minimal footprint of a region (8MB) and dictate the
+maximum size of stable memory for a canister today (32GB).  With 32k
+regions at 8MB each, well over the maximum stable memory size is used
+(256GB compared to 32GB, the limit today)
+
+
+## Design details
+
+### API
+
+Internal region allocator operations:
+
+
+ - `initialize` -- called by the RTS, not by the Motoko developer.
+ - `rebuild` -- also called by the RTS, during an post-upgrade.
+
+
+User-facing region allocator operations:
+
+ - `region_new` -- create a dynamic region.
+ - `region_grow` -- grow region by a specified number of pages.
+ - `region_release`* -- release region and reuse region's page blocks.
+ - `region_load` -- read some data from the region.
+ - `region_store` -- store some data into the region.
+
+
+ - * The `region_release` operation is not part of the MVP, but supporting it
+   is important because it means we can transition quickly to an integration
+   with the ambient Motoko GC if we can support it.
+
+## Internal footprint
+
+The state of the allocator is stored in a combination of stable memory and heap memory.
+
+The stable memory state is sufficient to reconstitute the heap objects
+(`rebuild` operation, described in a subsection below).  That means
+that even if the heap is lost (like during an upgrade that fails) the
+stable memory state can fully describe the state parts that should be
+rebuilt into the heap when the upgrade succeeds.
+
+### stable memory fields
+
+ - total allocated buckets, `Nat16`, max value is `32768`.
+ - total allocated regions, `Nat16`, max value is `32767` (one region is reserved for "no region").
+ - bucket-region table (fixed size, about 131kb).
+ - region-object table (fixed size, about 393kb).
+
+### bucket-region table
+
+ - purpose: relate a bucket ID ("page block ID") to its region (if any) and its position in that region.
+ - 32768 entries (statically sized).
+ - 4 bytes per entry.
+ - entry type = `BucketRegion { region : Nat16, position : Nat16 }`
+ - the location of each entry gives its corresponding bucket ID.
+
+### region-object table
+
+ - 32768 entries (statically sized).
+ - 12 bytes per entry.
+ - entry type = `RegionObjectEntry { size_in_pages: Nat64; heap_object: Nat32 }`
+ - the location of each entry gives its corresponding region ID.
+ - the heap object layout is rebuilt on upgrade.
+ - the heap object is described in its own subsection.
+
+### region heap object
+
+ - `RegionObject { id : Nat16; capacity : Nat16; vec_ptr: Nat32 }`
+ - 8 bytes per entry.
+ - `vec_ptr` points to a vector with `capacity` slots.
+ - to support load and store quickly, the `vec_ptr` is also held in the heap.
+ - capacity of `vec_ptr` doubles when it grows.
+ - no region has more than 32k page blocks, so a `Nat16` suffices for `capacity`.
+
+### Overview of `rebuild`
+
+When a canister upgrades, its heap is lost and the region heap objects must thus be rebuilt.
```

Comment:
>  if regions are stable, they will simply be saved into stable variables and preserved across upgrades in the usual way.

Your other comment about trying to avoid accesses to stable memory reminds me that a design here I had intended will be "too slow" in that sense.

The idea to represent a `Region` as a value was to use its ID.  Then, the operations would get the heap object by using that ID, which holds the vector of page blocks for the access.

Unfortunately, I seem to have neglected that first access (to get the heap object from its ID) when I accounted for how this would work fast enough.

If we ignore that, it seemed like a much simpler design than having a new kind of stable data variant for regions that get preserved in the heap directly (the ID would get preserved, and the heap object to which it referred would be `rebuilt`).  That simpler design avoids having to even serialize a new kind of `Region` value that lives in the heap (only its ID needs to be serialized, and that's easy).

Alas, though, it was not to be.  Thanks for catching this issue.

Reply:
> To arrange this, regions must be stable, and since they have state, like mutable arrays and fields, must be serialized in a way that preserves sharing (using the existing mechanism, but extended to region heap objects).

Hmm.  They do have mutable state, but that mutable state is very specialized.
 
For several reasons, I think it makes sense to treat that vector differently than a stable array.  It cannot be "shared" since it's allocated when a region object is allocated and that region object "owns" it.  Sharing regions does not change the reference count of the vector (still always just 1, the owning heap object.

If you ignore the page block vector's pointer, since it's not a user-level array and is hidden from user code, the region  has no "ordinary heap pointers" in it or that vector.  (Instead, it has sizes in the region object, and it has page block IDs in the vector).

Depending on what others think, I may still favor doing a `rebuild` step (sourced from those pre-allocated tables) on all of the stable regions' vectors, rather than try to serialize them in a more generic way, reusing some existing code.

So that is to say, the region object would be "stable", but its serialized form would null out the vector and its deserialization routine would rebuild it from the preallocated tables.

The space savings is trivial.  The value of not serializing the vectors seems mostly to be in greater simplicity over the alternative (and less code to go wrong, potentially).

Reply:
So I guess that leaves the issue of how to keep the heap values (growable segment vector and mutable page count) in sync with the disaster proof stable memory meta-data. Now we have two copies to maintain, and I guess for recoverability in case of a disaster, the stable memory version would need to be up-to-date. 
If you only save it on upgrade during serialization then I think you lose the disaster recovery.
You could save it at the end of every message, but then you'd need to keep tabs on those regions in a separate table somehow (otherwise you need to traverse the heap to find them all).
It might be worth asking the execution team when the perf optimizations for stable memory are due to come out. If reasonably soon (i.e. next month or so), then it might just be simpler to go with your original design and store the metadata in one place only - stable memory - and maintain it there, and just us a simple (user opaque) index to identify a region. That would also still be enough info to gc them, I think, if we ever do that.